Source generators that depend on other libraries/NuGet packages should have their dependencies passed as analyzers

# Problem Definition

I am currently in the process of converting a source generator I wrote in MSBuild to Bazel. The source generator depends on some first-party and some third-party packages.

This is very messy in MSBuild too, see https://github.com/dotnet/roslyn/discussions/47517.

Ideally, this would just work:

```starlark
# Core/BUILD.bazel
csharp_library(
    name = "Core",
    srcs = ["src/**/*.cs"],
    deps = [
        "@paket.main//some.package",
    ],
    target_frameworks = ["net8.0", "netstandard2.0"],
    langversion = "latest",
)

# SourceGenerator/BUILD.bazel
csharp_library(
    name = "SourceGenerator",
    is_analyzer = True,
    srcs = ["src/**/*.cs"],
    deps = [
        # First-party
        "//Core",
        # Third-party
        "@paket.main//microsoft.bcl.asyncinterfaces",
        "@paket.main//microsoft.codeanalysis.csharp",
    ],
    target_frameworks = ["netstandard2.0"],
    langversion = "latest",
)
```

However, I get the following error when I try this:

```
ERROR: /home/agni/[redacted]/Libraries/Foo.SourceGenerator.Tests/BUILD.bazel:9:18: Compiling Foo.SourceGenerator.Tests failed: (Exit 1): compiler_wrapper.sh failed: error executing CSharpCompile command (from target //Libraries/Foo.SourceGenerator.Tests:Foo.SourceGenerator.Tests) external/rules_dotnet~/dotnet/private/compiler_wrapper.sh external/rules_dotnet~~dotnet~dotnet_x86_64-unknown-linux-gnu/dotnet ... (remaining 2 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
warning CS8785: Generator 'FooSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'FileNotFoundException' with message 'Could not load file or assembly 'Microsoft.Bcl.AsyncInterfaces, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. The system cannot find the file specified.
'.
System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.Bcl.AsyncInterfaces, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. The system cannot find the file specified.

File name: 'Microsoft.Bcl.AsyncInterfaces, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'
   at Foo.SourceGenerator.FooSourceGenerator.Execute(SourceProductionContext context, ValueTuple`2 providers)
   at Microsoft.CodeAnalysis.UserFunctionExtensions.<>c__DisplayClass3_0`2.<WrapUserAction>b__0(TInput1 input1, TInput2 input2, CancellationToken token)
-----
```

# Solution

I see two possible solutions:
- Maybe some sort of `analyzer_deps` argument where dependencies to be passed as analyzer assemblies can be explicitly specified.
- If a first-party analyzer library is passed into `deps`, all of its dependencies should also be passed as analyzer assembles. I don't believe this violates `strict_deps` because we already have a well-defined dependency graph.

# Workarounds

For first-party packages, I can create a duplicate target that is `is_analyzer = True`, but for third-party packages no good solution exists (perhaps a custom Bazel rule that extracts the NuGetInfo from third-party packages and moves the libraries to `analyzers`, but we would still need to pass them manually).

--- a/dotnet/private/rules/csharp/actions/csharp_assembly.bzl
+++ b/dotnet/private/rules/csharp/actions/csharp_assembly.bzl
@@ -51,6 +51,24 @@ def _write_internals_visible_to_csharp(actions, label_name, dll_name, others):
 
     return output
 
+def _collect_analyzer_dependencies(deps):
+    """Collect the runtime libraries of this analyzer. These will be passed to the compiler.
+
+    Args:
+        deps: The list of dependencies of the analyzer.
+
+    Returns:
+        The list of analyzer dependencies.
+    """
+
+    dlls = []
+    for dep in deps:
+        runtime_info = dep[DotnetAssemblyRuntimeInfo]
+
+        # FIXME: Should this respect `not strict_deps`?
+        dlls.extend(runtime_info.libs)
+    return dlls
+
 # buildifier: disable=unnamed-macro
 def AssemblyAction(
         actions,
@@ -157,6 +175,12 @@ def AssemblyAction(
         exports,
         strict_deps,
     )
+
+    if (is_analyzer or is_language_specific_analyzer) and target_framework != "netstandard2.0":
+        fail("Analyzers must have `target_frameworks = [\"netstandard2.0\"]`.")
+
+    # TODO: Ensure that all the analyzer DLLs also target netstandard2.0.
+    analyzer_dlls = _collect_analyzer_dependencies(deps) if (is_analyzer or is_language_specific_analyzer) else []
 
     defines = framework_preprocessor_symbols(target_framework) + defines
 
@@ -303,8 +327,8 @@ def AssemblyAction(
         project_sdk = project_sdk,
         refs = [out_ref] if not is_analyzer else [],
         irefs = [out_iref] if out_iref else [out_ref],
-        analyzers = [] if (not is_analyzer) or is_language_specific_analyzer else [out_dll],
-        analyzers_csharp = [out_dll] if is_language_specific_analyzer else [],
+        analyzers = [] if (not is_analyzer) or is_language_specific_analyzer else ([out_dll] + analyzer_dlls),
+        analyzers_csharp = ([out_dll] + analyzer_dlls) if is_language_specific_analyzer else [],
         analyzers_fsharp = [],
         analyzers_vb = [],
         internals_visible_to = internals_visible_to or [],
@@ -319,7 +343,7 @@ def AssemblyAction(
     ), DotnetAssemblyRuntimeInfo(
         name = assembly_name,
         version = "1.0.0",  #TODO: Maybe make this configurable?
-        libs = [out_dll] if not is_analyzer else [],
+        libs = [out_dll] if not (is_analyzer or is_language_specific_analyzer) else [],
         pdbs = [out_pdb] if out_pdb else [],
         xml_docs = [out_xml] if out_xml else [],
         data = data,

--- a/dotnet/private/tests/analyzer_dependencies/Analyzer.cs
+++ b/dotnet/private/tests/analyzer_dependencies/Analyzer.cs
@@ -1,0 +1,5 @@
+namespace Analyzer;
+
+// NOTE: Doesn't actually do anything. Maybe the test could be enhanced by making this
+//       an actual analyzer.
+public static class Analyzer { }

--- a/dotnet/private/tests/analyzer_dependencies/BUILD.bazel
+++ b/dotnet/private/tests/analyzer_dependencies/BUILD.bazel
@@ -1,0 +1,3 @@
+load(":analyzer_dependencies.bzl", "test_analyzer_dependencies")
+
+test_analyzer_dependencies()

--- a/dotnet/private/tests/analyzer_dependencies/Core.cs
+++ b/dotnet/private/tests/analyzer_dependencies/Core.cs
@@ -1,0 +1,9 @@
+namespace Core;
+
+public class MyUtility
+{
+    public void Frobnicate()
+    {
+        System.Console.WriteLine("Frobnicating...");
+    }
+}

--- a/dotnet/private/tests/analyzer_dependencies/User.cs
+++ b/dotnet/private/tests/analyzer_dependencies/User.cs
@@ -1,0 +1,13 @@
+using Core;
+
+namespace User;
+
+public static class Program
+{
+    public static void Main()
+    {
+        // Ensure that the Core library was linked in as a runtime dependency.
+        var util = new MyUtility();
+        util.Frobnicate();
+    }
+}

--- a/dotnet/private/tests/analyzer_dependencies/analyzer_dependencies.bzl
+++ b/dotnet/private/tests/analyzer_dependencies/analyzer_dependencies.bzl
@@ -1,0 +1,107 @@
+"Tests for ensuring analyzers' dependencies are correctly passed in."
+
+load("@bazel_skylib//lib:unittest.bzl", "analysistest")
+load("//dotnet:defs.bzl", "csharp_library")
+load(
+    "//dotnet/private:providers.bzl",
+    "DotnetAssemblyCompileInfo",
+    "DotnetAssemblyRuntimeInfo",
+)
+
+def _force_netstandard20_transition_impl(_settings, _attr):
+    return {"//dotnet:target_framework": "netstandard2.0"}
+
+force_netstandard20_transition = transition(
+    implementation = _force_netstandard20_transition_impl,
+    inputs = [],
+    outputs = ["//dotnet:target_framework"],
+)
+
+def _has_analyzer_test_impl(ctx):
+    env = analysistest.begin(ctx)
+    target_under_test = analysistest.target_under_test(env)
+
+    compile_info = target_under_test[DotnetAssemblyCompileInfo]
+    if ctx.attr.target_language == "any":
+        actual_analyzer_libs = compile_info.analyzers + compile_info.transitive_analyzers.to_list()
+    elif ctx.attr.target_language == "csharp":
+        actual_analyzer_libs = compile_info.csharp_analyzers + compile_info.transitive_csharp_analyzers.to_list()
+    elif ctx.attr.target_language == "fsharp":
+        actual_analyzer_libs = compile_info.fsharp_analyzers + compile_info.transitive_fsharp_analyzers.to_list()
+    elif ctx.attr.target_language == "vb":
+        actual_analyzer_libs = compile_info.vb_analyzers + compile_info.transitive_vb_analyzers.to_list()
+    else:
+        fail("Unknown target language: {}".format(ctx.attr.target_language))
+
+    for expected_analyzer in ctx.attr.analyzers:
+        expected_compile_info = expected_analyzer[DotnetAssemblyCompileInfo]
+        expected_runtime_info = expected_analyzer[DotnetAssemblyRuntimeInfo]
+        if not expected_runtime_info.libs:
+            # This assembly is an analyzer itself.
+            expected_libs = expected_compile_info.analyzers
+        else:
+            # This assembly is not an analyzer, but its runtime libraries will be
+            # used by analyzers.
+            expected_libs = expected_runtime_info.libs
+
+        for expected_lib in expected_libs:
+            if expected_lib not in actual_analyzer_libs:
+                fail(
+                    "Expected analyzer library {} not found in: {}".format(
+                        expected_lib,
+                        actual_analyzer_libs,
+                    ),
+                )
+
+    return analysistest.end(env)
+
+has_analyzer_test = analysistest.make(
+    _has_analyzer_test_impl,
+    doc = "Tests whether the given target has an analyzer dependency.",
+    attrs = {
+        "analyzers": attr.label_list(
+            doc = "The list of analyzer dependencies to check for.",
+            mandatory = True,
+            providers = [DotnetAssemblyCompileInfo, DotnetAssemblyRuntimeInfo],
+            cfg = force_netstandard20_transition,
+        ),
+        "target_language": attr.string(
+            doc = "The target language to check for analyzers. Can be one of 'any', 'csharp', 'fsharp', or 'vb'.",
+            mandatory = True,
+            values = ["any", "csharp", "fsharp", "vb"],
+        ),
+    },
+)
+
+# buildifier: disable=function-docstring
+# buildifier: disable=unnamed-macro
+def test_analyzer_dependencies():
+    csharp_library(
+        name = "core_lib",
+        srcs = ["Core.cs"],
+        target_frameworks = ["net9.0", "netstandard2.0"],
+        langversion = "latest",
+    )
+
+    csharp_library(
+        name = "analyzer_lib",
+        is_analyzer = True,
+        srcs = ["Analyzer.cs"],
+        deps = [":core_lib"],
+        target_frameworks = ["netstandard2.0"],
+        langversion = "latest",
+    )
+
+    csharp_library(
+        name = "user_lib",
+        srcs = ["User.cs"],
+        deps = [":core_lib", ":analyzer_lib"],
+        target_frameworks = ["net9.0"],
+    )
+
+    has_analyzer_test(
+        name = "analyzer_dependency_was_passed_in",
+        target_under_test = ":user_lib",
+        analyzers = [":core_lib", ":analyzer_lib"],
+        target_language = "any",
+    )

